src/main/scala/strawman/collection/mutable/ArrayDeque.scala

```diff
@@ -0,0 +1,257 @@
+    checkIndex(idx)
+    set(idx, elem.asInstanceOf[AnyRef])
+  }
+
+  override def +=(elem: A) = {
+    ensureCapacity()
+    array(end) = elem.asInstanceOf[AnyRef]
+    end = (end + 1) & mask
+    this
+  }
+
+  override def +=:(elem: A) = {
+    ensureCapacity()
+    start = (start - 1) & mask
+    array(start) = elem.asInstanceOf[AnyRef]
+    this
+  }
+
+  override def ++=:(xs: TraversableOnce[A]) =
+    xs.foldRight(this)((x, coll) => x +=: coll).asInstanceOf[this.type]
+
+  override def insertAll(idx: Int, elems: scala.collection.Traversable[A]) = {
+    checkIndex(idx)
+    val src = elems.toBuffer
+    /*val finalLength = src.length + this.length
+    // Either we resize right away or move prefix right or suffix left
+    if (2*finalLength >= array.length) {
+      val array2 = ArrayDeque.alloc(finalLength)
+      arrayCopy(dest = array2, srcStart = 0, destStart = 0, maxItems = idx)
+      Array.copy(src = src, srcPos = 0, dest = array2, destPos = idx, length = src.length)
+      arrayCopy(dest = array2, srcStart = idx, destStart = idx + src.length, maxItems = size)
+      set(array = array2, start = 0, end = finalLength)
+    } else {*/
+    // TODO: choose to move prefix right or suffix left
+    val suffix = drop(idx)
+    end = (start + idx) & mask
+    this ++= src ++= suffix
+    /*}*/
+  }
+
+  override def remove(idx: Int, count: Int): Unit = {
+    checkIndex(idx)
+    if (count <= 0) return
+    val removals = (size - idx) min count
+    // If we are removing more than half the elements, its cheaper to start over
+    // Else, either move the prefix right or the suffix left - whichever is shorter
+    if(2*removals >= size) {
+      val array2 = ArrayDeque.alloc(size - removals)
+      arrayCopy(dest = array2, srcStart = 0, destStart = 0, maxItems = idx)
+      arrayCopy(dest = array2, srcStart = idx + removals - 1, destStart = idx, maxItems = size)
+      set(array = array2, start = 0, end = size - removals)
+    } else if (size - idx <= idx + removals) {
+      (idx until size) foreach {i =>
+        val elem = if (i + removals < size) get(i + removals) else null
+        set(i, elem)
+      }
+      end = (end - removals) & mask
+    } else {
+      (0 until (idx + removals)).reverse foreach {i =>
+        val elem = if (i - removals < 0) null else get(i - removals)
+        set(i, elem)
+      }
+      start = (start + removals) & mask
+    }
+  }
+
+  override def remove(idx: Int) = {
+    val elem = this(idx)
+    remove(idx, 1)
+    elem
+  }
+
+  def removeFirst(): Option[A] = {
+    if (isEmpty) {
+      None
+    } else {
+      val elem = array(start)
+      array(start) = null
+      start = (start + 1) & mask
+      Some(elem.asInstanceOf[A])
+    }
+  }
+
+  def removeLast(): Option[A] = {
+    if (isEmpty) {
+      None
+    } else {
+      end = (end - 1) & mask
+      val elem = array(end)
+      array(end) = null
+      Some(elem.asInstanceOf[A])
+    }
+  }
+
+  override def length = (end - start) & mask
```

Comment:
Yes, but I saw `size`, and when I found no definition I just deduced "it's from a superclass somehow" (I thought in `Buffer`).

But this is certainly a small thing, and disappears with an IDE.